### ISSUE TITLE
Compatability with dbtutils package.

### ISSUE THREAD

**Issue by KimchaC** [sqlfluff/sqlfluff#371]
Hi,

This is an amazing project and I am so excited to have stumbled upon it. My goal is to use this with DBT, but even after reading the documentation I am having trouble setting it up.

## My environment:
- dbt_utils package installed in `./dbt_modules/dbt_utils/macros`
- custom macros in `./macros/`

## If I run it with an empty config file:
```
$ sqlfluff lint models/01_source/is/src_is_other_tag_applications.sql                                                                            [15:03:45]
== [models/01_source/is/src_is_other_tag_applications.sql] FAIL
L:   - | P:   - |  TMP | Unrecoverable failure in Jinja templating:
                       | 'cast_is_datetime' is undefined. Have you configured
                       | your variables?
                       | https://docs.sqlfluff.com/en/latest/configuration.html
L:  10 | P:  12 |  TMP | Undefined jinja template variable: 'cast_is_datetime'
L:  11 | P:  12 |  TMP | Undefined jinja template variable: 'cast_as'
L:  12 | P:  12 |  TMP | Undefined jinja template variable: 'cast_as'
L:  27 | P:  12 |  TMP | Undefined jinja template variable: 'dbt_utils'
```

## When I add a config with the macros path:
```
[sqlfluff:templater:jinja]
load_macros_from_path=macros
```

Then I get...
```
$ sqlfluff lint models/01_source/is/src_is_other_tag_applications.sql                                                                            [15:16:35]
== [models/01_source/is/src_is_other_tag_applications.sql] FAIL
L:   - | P:   - |  TMP | Unrecoverable failure in Jinja templating: 'dbt_utils'
                       | is undefined. Have you configured your variables?
                       | https://docs.sqlfluff.com/en/latest/configuration.html
L:  27 | P:  12 |  TMP | Undefined jinja template variable: 'dbt_utils'
```

## When I add a config with the dbt_modules path:
```
[sqlfluff:templater:jinja]
load_macros_from_path=dbt_modules
```

Then I get an exception:
```
$ sqlfluff lint models/01_source/is/src_is_other_tag_applications.sql                                                                            [15:16:37]
Traceback (most recent call last):
  File "/Users/Kim/.pyenv/versions/3.8.2/bin/sqlfluff", line 11, in <module>
    load_entry_point('sqlfluff==0.3.4', 'console_scripts', 'sqlfluff')()
  File "/Users/Kim/.pyenv/versions/3.8.2/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/Kim/.pyenv/versions/3.8.2/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/Kim/.pyenv/versions/3.8.2/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/Kim/.pyenv/versions/3.8.2/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/Kim/.pyenv/versions/3.8.2/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/Kim/.pyenv/versions/3.8.2/lib/python3.8/site-packages/sqlfluff/cli/commands.py", line 162, in lint
    result = lnt.lint_paths(paths, verbosity=verbose, ignore_non_existent_files=False)
  File "/Users/Kim/.pyenv/versions/3.8.2/lib/python3.8/site-packages/sqlfluff/linter.py", line 890, in lint_paths
    result.add(self.lint_path(path, verbosity=verbosity, fix=fix,
  File "/Users/Kim/.pyenv/versions/3.8.2/lib/python3.8/site-packages/sqlfluff/linter.py", line 875, in lint_path
    self.lint_string(target_file.read(), fname=fname,
  File "/Users/Kim/.pyenv/versions/3.8.2/lib/python3.8/site-packages/sqlfluff/linter.py", line 692, in lint_string
    parsed, vs, time_dict = self.parse_string(s=s, fname=fname, verbosity=verbosity, config=config)
  File "/Users/Kim/.pyenv/versions/3.8.2/lib/python3.8/site-packages/sqlfluff/linter.py", line 590, in parse_string
    s, templater_violations = self.templater.process(s, fname=fname, config=config or self.config)
  File "/Users/Kim/.pyenv/versions/3.8.2/lib/python3.8/site-packages/sqlfluff/templaters.py", line 298, in process
    ctx.update(self._extract_macros_from_path(macros_path, env=env, ctx=live_context))
  File "/Users/Kim/.pyenv/versions/3.8.2/lib/python3.8/site-packages/sqlfluff/templaters.py", line 204, in _extract_macros_from_path
    macro_ctx.update(cls._extract_macros_from_path(
  File "/Users/Kim/.pyenv/versions/3.8.2/lib/python3.8/site-packages/sqlfluff/templaters.py", line 195, in _extract_macros_from_path
    cls._extract_macros_from_template(
  File "/Users/Kim/.pyenv/versions/3.8.2/lib/python3.8/site-packages/sqlfluff/templaters.py", line 173, in _extract_macros_from_template
    for k in macro_template.module.__dict__:
  File "/Users/Kim/.pyenv/versions/3.8.2/lib/python3.8/site-packages/jinja2/environment.py", line 1189, in module
    return self._get_default_module()
  File "/Users/Kim/.pyenv/versions/3.8.2/lib/python3.8/site-packages/jinja2/environment.py", line 1172, in _get_default_module
    self._module = rv = self.make_module()
  File "/Users/Kim/.pyenv/versions/3.8.2/lib/python3.8/site-packages/jinja2/environment.py", line 1155, in make_module
    return TemplateModule(self, self.new_context(vars, shared, locals))
  File "/Users/Kim/.pyenv/versions/3.8.2/lib/python3.8/site-packages/jinja2/environment.py", line 1238, in __init__
    body_stream = list(template.root_render_func(context))
  File "<template>", line 24, in root
  File "/Users/Kim/.pyenv/versions/3.8.2/lib/python3.8/site-packages/jinja2/sandbox.py", line 460, in call
    if not __self.is_safe_callable(__obj):
  File "/Users/Kim/.pyenv/versions/3.8.2/lib/python3.8/site-packages/jinja2/sandbox.py", line 360, in is_safe_callable
    getattr(obj, "unsafe_callable", False) or getattr(obj, "alters_data", False)
  File "/Users/Kim/.pyenv/versions/3.8.2/lib/python3.8/site-packages/jinja2/runtime.py", line 753, in __getattr__
    return self._fail_with_undefined_error()
  File "/Users/Kim/.pyenv/versions/3.8.2/lib/python3.8/site-packages/jinja2/runtime.py", line 747, in _fail_with_undefined_error
    raise self._undefined_exception(self._undefined_message)
jinja2.exceptions.UndefinedError: 'ref' is undefined
```

I have also tried adding the default macros to my own config, but the result was the same
```
[sqlfluff:templater:jinja]
load_macros_from_path=dbt_modules

[sqlfluff:templater:jinja:macros]
# Macros provided as builtins for dbt projects
dbt_ref = {% macro ref(model_ref) %}{{model_ref}}{% endmacro %}
dbt_source = {% macro source(source_name, table) %}{{source_name}}_{{table}}{% endmacro %}
dbt_config = {% macro config() %}{% for k in kwargs %}{% endfor %}{% endmacro %}
dbt_var = {% macro var(variable) %}item{% endmacro %}
```

## My questions
I think there may be a bug that causes the default macros to not be picked up.

And I am wondering whether it is possible to define multiple macro pathes. I think we need one for our custom macros path (./macros/) and one for the dbt_modules.

**Comment by cosoare** [comment]
Also stumbled upon this issue - not really sure how to setup sqlfluff to recognise `dbt_utils`
Can look at putting a fix if this is an issue that needs solving?

**Comment by aburgel** [comment]
I'm running into this issue as well. dbt packages are defined in a jinja variable/namespace (using [import](https://jinja.palletsprojects.com/en/2.11.x/templates/#import) it seems). I don't think there's a way to mirror this given what sqlfluff provides.

For a dbt template containing something like `{{ dbt_utils.surrogate_key('blah') }}`

I've tried:

```
[sqlfluff:templater:jinja:macros]
dbt_dbt_utils = {% macro dbt_utils() %}{% endmacro %}
dbt_surrogate_key = {% macro surrogate_key(keys) %}1{% endmacro %}
```

`dbt_utils` will be recognized but `surrogate_key` is not. And you can't define a macros as `dbt_utils.surrogate_key`

I also tried:

```
[sqlfluff:templater:jinja]
load_macros_from_path=sqlfluff_macros.sql
```

where `sqlfluff_macros.sql` contains:

```
{% import 'sqlfluff_dbt_utils.sql' as dbt_utils %}
```

and `sqlfluff_dbt_utils.sql` contains

```
{% macro surrogate_key(keys) %}1{% endmacro %}
```

That leads to this exception: `TypeError: no loader for this environment specified`

**Comment by justinstern** [comment]
I'm also running into this issue, while attempting to use `{{ dbt_utils.surrogate_key('blah') }}`:

```
$ sqlfluff lint models/*
== [models/staging/quickbooks/stg_quickbooks__purchase_lines.sql] FAIL
L:   - | P:   - |  TMP | Unrecoverable failure in Jinja templating: 'dbt_utils'
                       | is undefined. Have you configured your variables?
                       | https://docs.sqlfluff.com/en/latest/configuration.html
L:  13 | P:  12 |  TMP | Undefined jinja template variable: 'dbt_utils'
```

Tried pointing `.sqlfluff` at my dbt utils directory with: 

```
[sqlfluff:templater:jinja]
load_macros_from_path=dbt_modules/dbt_utils/macros
```

but that yields:

```
$ sqlfluff lint models/*

Traceback (most recent call last):
  File "/usr/local/bin/sqlfluff", line 8, in <module>
    sys.exit(cli())
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/sqlfluff/cli/commands.py", line 162, in lint
    result = lnt.lint_paths(paths, verbosity=verbose, ignore_non_existent_files=False)
  File "/usr/local/lib/python3.7/site-packages/sqlfluff/linter.py", line 891, in lint_paths
    ignore_non_existent_files=ignore_non_existent_files))
  File "/usr/local/lib/python3.7/site-packages/sqlfluff/linter.py", line 877, in lint_path
    fix=fix, config=config))
  File "/usr/local/lib/python3.7/site-packages/sqlfluff/linter.py", line 692, in lint_string
    parsed, vs, time_dict = self.parse_string(s=s, fname=fname, verbosity=verbosity, config=config)
  File "/usr/local/lib/python3.7/site-packages/sqlfluff/linter.py", line 590, in parse_string
    s, templater_violations = self.templater.process(s, fname=fname, config=config or self.config)
  File "/usr/local/lib/python3.7/site-packages/sqlfluff/templaters.py", line 298, in process
    ctx.update(self._extract_macros_from_path(macros_path, env=env, ctx=live_context))
  File "/usr/local/lib/python3.7/site-packages/sqlfluff/templaters.py", line 206, in _extract_macros_from_path
    env=env, ctx=ctx
  File "/usr/local/lib/python3.7/site-packages/sqlfluff/templaters.py", line 196, in _extract_macros_from_path
    template, env=env, ctx=ctx
  File "/usr/local/lib/python3.7/site-packages/sqlfluff/templaters.py", line 171, in _extract_macros_from_template
    macro_template = env.from_string(template, globals=ctx)
  File "/usr/local/lib/python3.7/site-packages/jinja2/environment.py", line 941, in from_string
    return cls.from_code(self, self.compile(source), globals, None)
  File "/usr/local/lib/python3.7/site-packages/jinja2/environment.py", line 638, in compile
    self.handle_exception(source=source_hint)
  File "/usr/local/lib/python3.7/site-packages/jinja2/environment.py", line 832, in handle_exception
    reraise(*rewrite_traceback_stack(source=source))
  File "/usr/local/lib/python3.7/site-packages/jinja2/_compat.py", line 28, in reraise
    raise value.with_traceback(tb)
  File "<unknown>", line 45, in template
jinja2.exceptions.TemplateSyntaxError: Encountered unknown tag 'materialization'.
```

<a name="workaround"></a>
## Workaround

Since the issue appears to be the `dbt_utils.` piece, I do have a bit of a workaround, though I don't love it because I'm basically redefining the `dbt_utils.surrogate_key` macro to remove the `dbt_utils` piece and will add a bit of a maintenance burden when I update dbt utils, should that macro change, but here it goes:

1. Copy the core dbt_utils `surrogate_key` macro file from `dbt_modules/dbt_utils/macros/sql/surrogate_key.sql` to (in my case): `macros/surrogate_key.sql`:

```
{%- macro surrogate_key() -%}

{% set fields = [] %}

{%- for field in varargs -%}

    {% set _ = fields.append(
        "coalesce(cast(" ~ field ~ " as " ~ dbt_utils.type_string() ~ "), '')"
    ) %}

    {% if not loop.last %}
        {% set _ = fields.append("'-'") %}
    {% endif %}

{%- endfor -%}

{{dbt_utils.hash(dbt_utils.concat(fields))}}

{%- endmacro -%}
```

2. Define `surrogate_key` as a macro within my `.sqlfluff` file:

```
[sqlfluff:templater:jinja:macros]
surrogate_key = {%- macro surrogate_key(name, type) -%}1{%- endmacro -%}
```

3. Within my model file, rather than using `{{ dbt_utils.surrogate_key('ID', 'value:Id::string') }}`, use the bare `{{ surrogate_key('ID', 'value:Id::string') }}`

And then everything is happy.

**Comment by alanmcruickshank** [comment]
Heya guys, thanks all for such detailed bug reports here. I'm aware that this is an issue.

Has anyone had this issue with any other plugins for dbt? Longer term I think sqlfluff needs a better way of rendering dbt queries that doesn't involve just duplicating the dbt plugin logic within sqlfluff.

This feels like one of the biggest issues at the moment which is affecting the most people so I'll try and find some time to work on this over the next few weeks. 👍 

Has anyone delved into the code enough have an suggestions on how they would fix this best? Very open to suggestions. If not I'll try a few things.

**Comment by dmateusp** [comment]
hey, I added a Draft PR as my "dive into the code" exercise, here's my approach:

* Support comma delimited list of directories in Configuration `load_macros_from_path`
* Support new comma delimited list of wildcard patterns to ignore in Configuration `load_macros_exclude_path_patterns`
* Check if macro sub-directories contain dbt_project.yml, in which case, import macros in dictionary `{<project_name>: ...}`

I can see a few issues with my approach already and I wanted to ask what people thought about optionally depending on the `dbt` project explicitly and using their code to read macros and detect sub-projects?

We started using the package at Earnest with DBT so I'm eager to find a solution to this

**Comment by pwildenhain** [comment]
I think we should consider letting `dbt` do the compiling: https://github.com/sqlfluff/sqlfluff/issues/381#issuecomment-703138462

> 
> 
> > @alanmcruickshank: I'm considering changing how we support the compiling of templates and actually getting dbt to do the rendering of queries, which would solve a load of these issues.
> 
> @pwildenhain: I'm 💯 in agreement with this. If people are using `dbt` _anyway_ with this project, then I think we should just use `dbt` to compile the queries, and eliminate the need to double document macros
> 
> Could this be added as a config option? Should this be a separate plugin (like `sqlfluff-dbt`)?

**Comment by huntdm07** [comment]
I'm also hitting this issue - I'd say it's the one roadblock preventing me from rolling this out to our team, since we use `dbt_utils.surrogate_key` or `dbt_utils.group_by` in almost every model. Hoping that a fix might mean we don't need to duplicate the macros across to our project (although it's not the worst fallback for just a few functions).

**Comment by pwildenhain** [comment]
Yup, we think we might have a fix with the work being done in #508, keep an eye on that conversation 👀 to stay in the loop

**Comment by alanmcruickshank** [comment]
I'm going to suggest that this is best achieved with the new `dbt` templater built into sqlfluff.

For people watching this issue (@KimchaC , @cosoare , @aburgel , @justinstern , @huntdm07 , @jimmyocto , @tnightengale, @danielsepulvedab ), there is now an alpha release of sqlfluff 0.4.0 available on pypi (https://pypi.org/project/sqlfluff/0.4.0a1/).

I would be very grateful for feedback on the work that @dmateusp has done to get us this far as part of #508 .

In particular you can find docs for the new dbt templater here: https://docs.sqlfluff.com/en/latest/configuration.html#dbt-project-configuration

It would be useful to know whether:
a) the feature satisfies all the requirements here
b) whether the current implementation works in production
b) whether the documentation that we've got is sufficient for users to set things up

**Comment by KimchaC** [comment]
Guys, you have done an absolutely amazing job with this. I followed the documentation and was able to get it running in no time and it is working well with my rather large project.

Here are a few things I ran into:

## Difficulty matching line numbers
This is probably a tough problem to solve, but the line numbers sqlfluff spits out are of course the line numbers of the compiled SQL and not of the jinja SQL file. So it is hard to find where the error is without looking at the compiled file.

This is going to make it difficult to use the linter in a code editor.

Perhaps one solution would be to have a mode where instead of linting the compile code, you ignore the jinja statements and just lint the SQL code around it.

Or perhaps you lint whether the jinja code is valid jinja and compiles well, plus the raw SQL around it.

This also brings me to...

## I couldn't get the VSCode extension working
I have enabled the extension and added this to my vscode settings.json:
```
{
    "sql.linter.run": "onSave",
    "sql.linter.executablePath": "/Users/Kim/.pyenv/shims/sqlfluff"
}
```

But nothing is being run.

## {% raw %} causes an error

I have SQL code like this...
```
        {% raw %}
        final_url_replaced = replace(replace(replace(
            final_url_parsed, 
            '{campaignid}', cast(campaign_id as varchar(25)) ),
            '{adgroupid}', cast(ad_group_id as varchar(25)) ), 
            '{creative}', cast(ad_id as varchar(25))
        ),
        {% endraw %}
```

And sqlfluff throws this error:

```
Traceback (most recent call last):
  File "/Users/Kim/.pyenv/versions/3.8.6/bin/sqlfluff", line 8, in <module>
    sys.exit(cli())
  File "/Users/Kim/.pyenv/versions/3.8.6/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/Kim/.pyenv/versions/3.8.6/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/Kim/.pyenv/versions/3.8.6/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/Kim/.pyenv/versions/3.8.6/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/Kim/.pyenv/versions/3.8.6/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/Kim/.pyenv/versions/3.8.6/lib/python3.8/site-packages/sqlfluff/cli/commands.py", line 320, in lint
    result = lnt.lint_paths(
  File "/Users/Kim/.pyenv/versions/3.8.6/lib/python3.8/site-packages/sqlfluff/core/linter.py", line 1154, in lint_paths
    self.lint_path(
  File "/Users/Kim/.pyenv/versions/3.8.6/lib/python3.8/site-packages/sqlfluff/core/linter.py", line 1135, in lint_path
    self.lint_string(
  File "/Users/Kim/.pyenv/versions/3.8.6/lib/python3.8/site-packages/sqlfluff/core/linter.py", line 936, in lint_string
    parsed = self.parse_string(in_str=in_str, fname=fname, config=config)
  File "/Users/Kim/.pyenv/versions/3.8.6/lib/python3.8/site-packages/sqlfluff/core/linter.py", line 704, in parse_string
    templated_file, templater_violations = self.templater.process(
  File "/Users/Kim/.pyenv/versions/3.8.6/lib/python3.8/site-packages/sqlfluff/core/templaters/dbt.py", line 181, in process
    return self._unsafe_process(fname, in_str, config)
  File "/Users/Kim/.pyenv/versions/3.8.6/lib/python3.8/site-packages/sqlfluff/core/templaters/dbt.py", line 227, in _unsafe_process
    raw_sliced, sliced_file = self.slice_file(node.raw_sql, node.injected_sql)
  File "/Users/Kim/.pyenv/versions/3.8.6/lib/python3.8/site-packages/sqlfluff/core/templaters/python.py", line 254, in slice_file
    raw_sliced = list(cls._slice_template(raw_str))
  File "/Users/Kim/.pyenv/versions/3.8.6/lib/python3.8/site-packages/sqlfluff/core/templaters/jinja.py", line 312, in _slice_template
    block_type = block_types[elem_type]
KeyError: 'raw_end'
```

## .sqlfluffignore doesn't work when adding a path to the `lint` command
I added the files using `{% raw %} ` to my .sqlfluffignore...

```
# Models using {% raw %}
models/02_base/base_fb_ads.sql
models/02_base/base_google_ad_insights.sql
models/02_base/base_google_ads.sql
```

And when I run `sqlfluff lint .` in the same directory as the .sqlfluffignore, it works and the files are ignored. But when I run `sqlfluff lint models` they are not ignored.

## No MS SQL Server Dialect
It's a pretty popular DB so it would be great if you could add support for it too.

## DBT Dependency
One thing that would be great is if we could configure the path to DBT instead of making it a requirement in pip. For example, on my mac I had DBT installed using brew, but to make sqlfluff work I had to install it again.

No big deal, but just something to consider.

**Comment by NiallRees** [comment]
Hey @KimchaC thank you so so much for the testing on this! 🙇‍♂️

I'm going to close this original issue as it has been resolved by #508, but please raise any issues you've encountered separately, where they'll be far more visible too.

**Comment by Felix-Blom-SV** [comment]
@justinstern A much easier way is the following:

Create macro, simply wrap the function:
```sql
{% macro dbt_utils_generate_surrogate_key(columns) %}
{{ dbt_utils.generate_surrogate_key(columns) }}
{% endmacro %}
```

In the `.sqlfluff` file/same thing you did:
```
[sqlfluff:templater:jinja:macros]
dbt_utils_generate_surrogate_key = {% macro dbt_utils_generate_surrogate_key(columns) %}'surrogate_key'{% endmacro %}
```

I'm using this because I don't want a db connection needed every time I lint (I want to use jinja not dbt)
